<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Guide · GenomeGraphs.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomeGraphs.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Manual</span><ul><li class="current"><a class="toctext" href>Guide</a><ul class="internal"><li><a class="toctext" href="#Installation-1">Installation</a></li><li><a class="toctext" href="#Creating-a-WorkSpace-1">Creating a WorkSpace</a></li><li><a class="toctext" href="#Constructing-a-de-Bruijn-graph-1">Constructing a de-Bruijn graph</a></li></ul></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>Guide</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/man/guide.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Guide</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Package-Guide-1" href="#Package-Guide-1">Package Guide</a></h1><h2><a class="nav-anchor" id="Installation-1" href="#Installation-1">Installation</a></h2><p>GenomeGraphs is made available to install through BioJulia&#39;s package registry.</p><p>Julia by default only watches the &quot;General&quot; package registry, so before you start, you should add the BioJulia package registry.</p><p>Start a julia terminal, hit the <code>]</code> key to enter pkg mode (you should see the prompt change from <code>julia&gt;</code> to <code>pkg&gt;</code>), then enter the following command:</p><pre><code class="language-none">pkg&gt; registry add https://github.com/BioJulia/BioJuliaRegistry.git</code></pre><p>After you&#39;ve added the registry, you can install GenomeGraphs from the julia REPL. Press <code>]</code> to enter pkg mode again, and enter the following:</p><pre><code class="language-none">pkg&gt; add GenomeGraphs</code></pre><p>If you are interested in the cutting edge of the development, please check out the master branch to try new features before release.</p><h2><a class="nav-anchor" id="Creating-a-WorkSpace-1" href="#Creating-a-WorkSpace-1">Creating a WorkSpace</a></h2><p>You can create an empty genome graph workspace with the empty constructor:</p><pre><code class="language-julia">using GenomeGraphs

ws = WorkSpace()</code></pre><pre><code class="language-none"> Sequence distance graph (0 nodes)
</code></pre><p>Obviously this workspace is quite useless on its own!</p><p>You need a genome graph and information to project onto it before you can do any exploration or analysis. There are a few ways to get your first graph:</p><ol><li>Construct a de-Bruijn graph from raw sequencing reads.</li><li>Load a graph (such as produced from another assembler) from a GFAv1 file</li></ol><h2><a class="nav-anchor" id="Constructing-a-de-Bruijn-graph-1" href="#Constructing-a-de-Bruijn-graph-1">Constructing a de-Bruijn graph</a></h2><p>Let&#39;s see how to do option number 1, and construct a de-Bruijn graph from raw sequencing reads. This can be achieved with a few simple steps:</p><ol><li>Prepare the sequencing reads.</li></ol><h3><a class="nav-anchor" id="Preparing-the-sequencing-reads-1" href="#Preparing-the-sequencing-reads-1">Preparing the sequencing reads</a></h3><p>Let&#39;s prepare the sequencing reads.</p><p><code>WorkSpaces</code> store sequencing reads in <code>ReadDatastore</code>s, provided by the <a href="https://github.com/BioJulia/ReadDatastores.jl">ReadDatastores.jl</a> package.</p><p>GenomeGraphs uses (and re-exports types and methods from) <code>ReadDatastores</code>. <code>ReadDatastores</code> is a standalone package in its own right (although it was built in the first place for GenomeGraphs).</p><p>If you want to use <code>ReadDatastores</code> as a standalone package in another Bio(Julia) based project (and we recommend you do - the data stores are more efficient than text files), you can find standalone docs for the package <a href="https://biojulia.net/ReadDatastores.jl/latest/">here</a>. Some types and methods documented there, are repeated in the Library section of this manual here, for convenience.</p><p>Anyway, let&#39;s build our first paired end reads datastore!</p><pre><code class="language-">using GenomeGraphs

fwq = open(FASTQ.Reader, &quot;test/ecoli_pe_R1.fastq&quot;)
rvq = open(FASTQ.Reader, &quot;test/ecoli_pe_R2.fastq&quot;)

ds = PairedReads(fwq, rvq, &quot;ecoli-test-paired&quot;, &quot;my-ecoli&quot;, 250, 300, 0, FwRv)</code></pre><p>Here I gave my datastore the base filename of &quot;ecoli-test-paired&quot;, and gave the datastore a name of &quot;my-ecoli&quot;. I set a minimum length for the reads at 250 base pairs, and the maximum length to 300 base pairs. Reads that are too short are discarded, reads that are too long are truncated.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>I set the insert size of the paired reads to 0, since I&#39;m not sure of it and right now the value is optional.</p></div></div><p>I set the orientation of the paired reads to <code>FwRv</code>. This is the default, and means for every pair of reads, read 1 is oriented in the forward direction, and read 2 is oriented backwards (forwards on the opposite strand). This orientation distinguishes regular paired-end reads from other paired read types like Long Mate Pairs.</p><p>Now the datastore is created, it can be added to a workspace.</p><pre><code class="language-">ws = WorkSpace()
add_paired_reads!(ws, ds)</code></pre><h3><a class="nav-anchor" id="Run-the-dbg-process-1" href="#Run-the-dbg-process-1">Run the <code>dbg</code> process</a></h3><p>GenomeGraphs comes with some very high-level methods in it&#39;s API, that we like to call <em>processes</em>. The perform some critical and common task as part of a larger workflow. Examples include constructing a de-Bruijn graph from sequencing reads, mapping reads to a graph, kmer counting and so on.</p><p>Once a workspace has an attached read datastore, you can run the dbg process to produce a first de-Bruijn graph of the genome.</p><pre><code class="language-">dbg!(BigDNAMer{61}, 10, ws, &quot;my-ecoli&quot;)</code></pre><footer><hr/><a class="previous" href="../../"><span class="direction">Previous</span><span class="title">Home</span></a></footer></article></body></html>
