<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Graphs submodule · GenomeGraphs.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="GenomeGraphs.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">GenomeGraphs.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/guide/">Guide</a></li></ul></li><li><span class="tocitem">API</span><ul><li class="is-active"><a class="tocitem" href>Graphs submodule</a><ul class="internal"><li><a class="tocitem" href="#Types-1"><span>Types</span></a></li><li><a class="tocitem" href="#Public-/-Safe-methods-1"><span>Public / Safe methods</span></a></li><li><a class="tocitem" href="#Internal-/-Unsafe-methods-1"><span>Internal / Unsafe methods</span></a></li></ul></li><li><a class="tocitem" href="../MerTools/">MerTools submodule</a></li><li><a class="tocitem" href="../GraphIndexes/">GraphIndexes submodule</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Graphs submodule</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Graphs submodule</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/api/Graphs.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API:-The-Graphs-submodule-1"><a class="docs-heading-anchor" href="#API:-The-Graphs-submodule-1">API: The Graphs submodule</a><a class="docs-heading-anchor-permalink" href="#API:-The-Graphs-submodule-1" title="Permalink"></a></h1><p>The Graphs submodule contains the core graph data-structure used in GenomeGraphs.jl: The SequenceDistanceGraph.</p><p>It also defines some of the common methods used in the rest of the higher-level parts GenomeGraphs.jl:</p><ul><li>To query nodes/sequences in the graph.</li><li>To find certain topological structures.</li><li>To edit graph structure.</li><li>To write/load SequenceDistanceGraph&#39;s to/from file.</li></ul><p>Some methods, particularly ones that access the sequences stored in the graph, and to edit topology, are marked as unsafe and are not exported deliberately.</p><p>All types and methods defined in this module are documented here.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This is a reference of an internal sub-module&#39;s API for developers and experienced users. First ask yourself if what you need isn&#39;t covered by the higher-level WorkSpace API. </p></div></div><h2 id="Types-1"><a class="docs-heading-anchor" href="#Types-1">Types</a><a class="docs-heading-anchor-permalink" href="#Types-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.SequenceDistanceGraph" href="#GenomeGraphs.Graphs.SequenceDistanceGraph"><code>GenomeGraphs.Graphs.SequenceDistanceGraph</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The SequenceDistanceGraph is a representation of a genome assembly.</p><p>The SequenceDistanceGraph stores sequences in a vector of <code>SequenceDistanceGraphNode</code>s (or <code>SDGNode</code> for short). </p><p>A singe node represents a sequence <em>and</em> its reverse complement.</p><p>Every node also has an orientaton: Each node has a positive end (+), and a negative end (-). So when a node is accessed with (or traversed by entering) the positive end the node yields the stored sequence. Conversely, when a node is accessed with (or traversed by entering) the negative end the node yelds the reverse complement of the stored sequence. In this way the positive end can be thought of as the sequence start of the node&#39;s canonical sequence, and the negative end can be thought of as its end.</p><p>Thankfully, you don&#39;t need to think about &quot;positive node ends&quot; and &quot;negative  node ends&quot; most of the time you interact with a <code>SequenceDistanceGraph</code>: In a <code>SequenceDistanceGraph</code>, every has a correlative ID starting from 1. For every node stored in the graph with an ID X, the negative ID -X is mapped to the reverse complement of that node. This mapping is virtual: Only one node is stored in the graph. To see more about how these positive and negative node IDs work, refer to the docs for methods such as <code>get_next_nodes</code> and <code>get_previous_nodes</code>. These high level functions simplify coding graph traversal A LOT.</p><p>This <code>SequenceDistanceGraph</code> type is only intended to be interacted with directly by developers and users who know what they are doing. Most tasks an end user wants to do to manipulate or query a graph can be achieved through the  <code>WorkSpace</code> type and the NodeView interface.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.SDG" href="#GenomeGraphs.Graphs.SDG"><code>GenomeGraphs.Graphs.SDG</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Shorthand for SequenceDistanceGraph</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.SDGNode" href="#GenomeGraphs.Graphs.SDGNode"><code>GenomeGraphs.Graphs.SDGNode</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The SDGNode type represents a node in a SequenceDistanceGraph.</p><p>At present it contains only two fields, first it holds an instance of a BioSequences BioSequence type.</p><p>Secondly, it tracks a flag which indicates if the node has been deleted or not.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The deleted flag allows us to mark nodes in the graph as deleted, which can be of help in some algorithms where the graph structure is being edited (merging nodes for example).</p><p>Actually deleting the node would shift node IDs and require redoing all links in the graph and so on.</p><p>So just marking a node as deleted and not using it anymore is a lazy but sometimes helpful choice.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.SequenceDistanceGraphLink" href="#GenomeGraphs.Graphs.SequenceDistanceGraphLink"><code>GenomeGraphs.Graphs.SequenceDistanceGraphLink</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A SequenceDistanceGraphLink or SDGLink represents a single distance between two sequences in a SequenceDistanceGraph.</p><p>Every link describes a connection between two node ends and contains a distance (they take the form <code>([+, -]n1, [+, -]n2, [+, -]dist)</code>).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>A link connects two node ends, and so the order of the signed nodes in the links does not change the link.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>If the distance in a link is negative, this represents an overlap between two sequences. These overlaps must be <strong>perfect overlaps</strong>.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Two links are considered equal (<code>==</code> will return <code>true</code>) if the source and the destination fields of the two links are the same, the distance or dist field is not considered!</p></div></div></div></section></article><h2 id="Public-/-Safe-methods-1"><a class="docs-heading-anchor" href="#Public-/-Safe-methods-1">Public / Safe methods</a><a class="docs-heading-anchor-permalink" href="#Public-/-Safe-methods-1" title="Permalink"></a></h2><h3 id="Graph-nodes-and-sequences-1"><a class="docs-heading-anchor" href="#Graph-nodes-and-sequences-1">Graph nodes and sequences</a><a class="docs-heading-anchor-permalink" href="#Graph-nodes-and-sequences-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.name" href="#GenomeGraphs.Graphs.name"><code>GenomeGraphs.Graphs.name</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the name of the graph. Defaults to the symbol :sdg.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.n_nodes" href="#GenomeGraphs.Graphs.n_nodes"><code>GenomeGraphs.Graphs.n_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the number of nodes in the sequence distance graph <code>sg</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.sequence" href="#GenomeGraphs.Graphs.sequence"><code>GenomeGraphs.Graphs.sequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">sequence(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the full sequence of a node in a sequence distance graph using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>sequence</code> accepts a NodeID that can be positive or negative. Nodes represent stretches of sequence in a canonical orientation, if you ask for for the sequence of say the third node, the positive node id 3 (which denotes traversing the third node in the forward direction), gives you the canonical sequence. If you use the negative ID -3 (which denotes traversing the third node in the reverse direction), you will get the reverse complement of the node&#39;s canonical (forward) sequence.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>It is safe to modify the returned sequence without screwing up your graph, yet thanks to BioSequences.jl&#39;s copy on write system for LongSequences, data copying will only occur if nessecery. You get the best of both worlds.</p></div></div></div></section></article><h3 id="Graph-Topology-1"><a class="docs-heading-anchor" href="#Graph-Topology-1">Graph Topology</a><a class="docs-heading-anchor-permalink" href="#Graph-Topology-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.source" href="#GenomeGraphs.Graphs.source"><code>GenomeGraphs.Graphs.source</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the source node end of a link.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.destination" href="#GenomeGraphs.Graphs.destination"><code>GenomeGraphs.Graphs.destination</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the destination node end of a link.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.distance" href="#GenomeGraphs.Graphs.distance"><code>GenomeGraphs.Graphs.distance</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the distance between two node ends involved in a link.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.is_forwards_from" href="#GenomeGraphs.Graphs.is_forwards_from"><code>GenomeGraphs.Graphs.is_forwards_from</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Test if link <code>l</code> is a forward link leaving node <code>n</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.is_backwards_from" href="#GenomeGraphs.Graphs.is_backwards_from"><code>GenomeGraphs.Graphs.is_backwards_from</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Test if link <code>l</code> is a forward link leaving node <code>n</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.find_link" href="#GenomeGraphs.Graphs.find_link"><code>GenomeGraphs.Graphs.find_link</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_link(sg::SequenceDistanceGraph, src::NodeID, dst::NodeID)</code></pre><p>Find and return the link that exists between a source node and a destination node, using their correlative node ids.</p><p>In this instance, the IDs also denote the &quot;ends&quot; of a node, a link connects.</p><p>Recall that every node has an orientaton: Each node has a positive end (+), and a negative end (-). So when a node is accessed with (or traversed by entering) the positive end the node yields the stored sequence. Conversely, when a node is accessed with (or traversed by entering) the negative end the node yelds the reverse complement of the stored sequence.</p><p>Thus <code>find_link(sg, -5, 1)</code> means you want to find a link in the graph that allows you to exit the (-) end of node 5 (meaning you just traversed it in the  canonical + –&gt; - orientation), and enter the (+) end of node 1 (meaning you will also traverse node one in the canonical + –&gt; - orientation).</p><p>By contrast <code>find_link(sg, -5, -1)</code> means you want to find a link in the graph that allows you to exit the (-) end of node 5 - again having just traverse it in the + –&gt; - orientation, but want to enter node 1 through it&#39;s (-) end - and thus traversing node 1 in the + &lt;– - orientation, yielding the reverse complement of the canonical sequence of node 1.</p><p>If a link connecting the two node ends is present, you get it. If not, you get <code>nothing</code>. So make sure to check the output.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method is safe and public, but not reeeeaaallly intended for the end user, as they have to worry about links between &quot;ends&quot; of a node. In contrast to the rest of this framework, which rather has the user thinking in node centric terms. e.g. &quot;I am at node 5 which means I have traversed it forwards or I am at node -5, which means I have traversed node 5 backwards. From here I can visit node 1, -3, and 2, so I may go through nodes 1 and 2 forwards, and through 3 backwards. The methods <code>get_next_nodes</code> and <code>get_previous_nodes</code> are more user-friendly and follow this more node-centric mindset.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.forward_links" href="#GenomeGraphs.Graphs.forward_links"><code>GenomeGraphs.Graphs.forward_links</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">forward_links(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a vector of the links that are ahread of you, as you traverse node <code>n</code>, continuing forward in you present direction of travel (<code>n</code> can be a positive or negative node id).</p><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the links that allow you to leave node 5 in the canonical direction. If you used a negative ID such as -5, that means you are traversing node 5 in the non-canonical orientation, and so you will get the links that allow you to leave node 5 in the non-canonical direction.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.backward_links" href="#GenomeGraphs.Graphs.backward_links"><code>GenomeGraphs.Graphs.backward_links</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">backward_links(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a vector of the links that are behind of you, as you traverse node <code>n</code>, continuing forward in you present direction of travel (<code>n</code> can be a positive or negative node id).</p><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the links that would allow you to enter node 5 in the canonical direction. If you used a negative ID such as -5, that means you are traversing node 5 in the non-canonical orientation, and so you will get the links that would allow you to enter node 5 in the non-canonical direction.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.get_next_nodes" href="#GenomeGraphs.Graphs.get_next_nodes"><code>GenomeGraphs.Graphs.get_next_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_next_nodes(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the nodes you may visit next as you exit node <code>n</code>, maintaining your current direction of travel (<code>n</code> can be a positive or negative node ID).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the nodes you may visit as you leave node 5 in the canonical direction. If you use a negative ID, such as -5, this means you are traversing node 5 in the non-canonical orientation, and so you will get the nodes you may visit as you leave node 5 in the non-canonical direction.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The list of nodes returned are also signed to denote direction: Say you got <code>[1, -2, 3]</code> as a result of <code>get_next_nodes(sg, 5)</code>. That means as you leave node 5 after traversing it in the canonical direction, you may proceed to travel through nodes 1 and 3 in the canonical direction, and node 2 in the non-canonical direction.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.get_previous_nodes" href="#GenomeGraphs.Graphs.get_previous_nodes"><code>GenomeGraphs.Graphs.get_previous_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_previous_nodes(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the nodes you may have previously been on before you enterd node <code>n</code>, maintaining your current direction of travel (<code>n</code> can be a positive or negative node ID).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the nodes you may have visited prior to entering node 5 in the canonical direction. If you use a negative ID, such as -5, this means you are traversing node 5 in the non-canonical orientation, and so you will get the nodes you may have visited prior to entering node 5 in the non-canonical direction.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The list of nodes returned are also signed to denote direction: Say you got <code>[1, -2, 3]</code> as a result of <code>get_previous_nodes(sg, 5)</code>. That means prior to entering node 5 in the canonical direction, you may have traveled through nodes 1 and 3 in the canonical direction, and node 2 in the non-canonical direction.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.find_tip_nodes" href="#GenomeGraphs.Graphs.find_tip_nodes"><code>GenomeGraphs.Graphs.find_tip_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_tip_nodes(sg::SequenceDistanceGraph, min_size::Integer)</code></pre><p>Get a set of IDs of all the nodes in the graph <code>sg</code> that count as &quot;tips&quot;.</p><p>Here a tip node is defined as a node that only has a single neighbouring node at one of it&#39;s ends, and no neighbouring nodes out of one of its other end.</p><p>The node&#39;s sequence must also be larger than <code>min_size</code> base pairs in length.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.find_tip_nodes!" href="#GenomeGraphs.Graphs.find_tip_nodes!"><code>GenomeGraphs.Graphs.find_tip_nodes!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_tip_nodes!(result::Set{NodeID}, sg::SequenceDistanceGraph, min_size::Integer)</code></pre><p>Get a set of IDs of all the nodes in the graph <code>sg</code> that count as &quot;tips&quot;.</p><p>Here a tip node is defined as a node that only has a single neighbouring node at one of it&#39;s ends, and no neighbouring nodes out of one of its other end.</p><p>The node&#39;s sequence must also be larger than <code>min_size</code> base pairs in length.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method modifys a <code>result</code> input. So if you want to find tips in a graph repeatedly, you can resuse a <code>Set</code>, saving you some allocation costs.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.find_all_unitigs" href="#GenomeGraphs.Graphs.find_all_unitigs"><code>GenomeGraphs.Graphs.find_all_unitigs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_all_unitigs(sg::G, min_nodes::Integer) where {G&lt;:SequenceDistanceGraph}</code></pre><p>Find and return a vector of paths through the graph that represent all the unitigs or transitive paths in the graphs. Such paths are defined as a chain of nodes with only one neighbour. Such simple regions of the graph can safely be collapsed into one larger node.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.find_all_unitigs!" href="#GenomeGraphs.Graphs.find_all_unitigs!"><code>GenomeGraphs.Graphs.find_all_unitigs!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_all_unitigs!(unitigs::Vector{SequenceDistanceGraphPath{G}}, sg::G, min_nodes::Integer) where {G&lt;:SequenceDistanceGraph}</code></pre><p>Find and return a vector of paths through the graph that represent all the unitigs or transitive paths in the graphs. Such paths are defined as a chain of nodes with only one neighbour. Such simple regions of the graph can safely be collapsed into one larger node.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This vector modifies an input <code>unitigs</code> vector to contain the result. This is useful for situations where you want to repeatedly find unitigs in the graph to save on additional allocations.</p></div></div></div></section></article><h3 id="Graph-IO-1"><a class="docs-heading-anchor" href="#Graph-IO-1">Graph IO</a><a class="docs-heading-anchor-permalink" href="#Graph-IO-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.write_to_gfa1" href="#GenomeGraphs.Graphs.write_to_gfa1"><code>GenomeGraphs.Graphs.write_to_gfa1</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">write_to_gfa1(sg::SequenceDistanceGraph, filename::String)</code></pre><p>Write a graph from a GFAv1 formatted file, and associated FASTA file of node sequences.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The GFA format permits storing sequences of the graph nodes in a seperate fasta file, instead of in the GFA file. This is so as the sequences of the graph nodes can be easily fed into other tools that typically accept FASTA files as input. Many assemblers also output a GFA + FASTA combo.  Therefore, this method writes node sequences to a file called <code>[filename].fasta</code>, and the graph structure to a file called <code>[filename].gfa</code>.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.load_from_gfa1!" href="#GenomeGraphs.Graphs.load_from_gfa1!"><code>GenomeGraphs.Graphs.load_from_gfa1!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">load_from_gfa1!(sg::SequenceDistanceGraph{S}, gfafile::AbstractString, fafile::AbstractString) where {S&lt;:BioSequence}</code></pre><p>Load a graph from a GFAv1 formatted file, and associated FASTA file of node sequences.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The GFA format permits storing sequences of the graph nodes in a seperate fasta file, instead of in the GFA file. This is so as the sequences of the graph nodes can be easily fed into other tools that typically accept FASTA files as input. Many assemblers also output a GFA + FASTA combo.  Therefore, this method asks for the filepath of a GFAv1 file, as well as a filepath to a FASTA formatted file. This method reads the node sequences from the FASTA file, before getting the links between nodes from the GFAv1 file.   </p></div></div></div></section></article><h2 id="Internal-/-Unsafe-methods-1"><a class="docs-heading-anchor" href="#Internal-/-Unsafe-methods-1">Internal / Unsafe methods</a><a class="docs-heading-anchor-permalink" href="#Internal-/-Unsafe-methods-1" title="Permalink"></a></h2><h3 id="Graph-nodes-and-sequences-2"><a class="docs-heading-anchor" href="#Graph-nodes-and-sequences-2">Graph nodes and sequences</a><a class="docs-heading-anchor-permalink" href="#Graph-nodes-and-sequences-2" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.empty_seq" href="#GenomeGraphs.Graphs.empty_seq"><code>GenomeGraphs.Graphs.empty_seq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">empty_seq(::Type{LongSequence{DNAAlphabet{2}}})</code></pre><p>Get a reference to an empty 2-bit long dna sequence.</p><p>In this module, a single empty sequence is allocated, and every call to <code>empty_seq</code> returns a reference to that same allocated sequence.</p><p>This saves on allocations if you are deleting many nodes in a SequenceDistanceGraph.</p></div></section><section><div><p>empty_seq(::Type{LongSequence{DNAAlphabet{4}}})</p><p>Get a reference to an empty 4-bit long dna sequence.</p><p>In this module, a single empty sequence is allocated, and every call to <code>empty_seq</code> returns a reference to that same allocated sequence.</p><p>This saves on allocations if you are deleting many nodes in a SequenceDistanceGraph.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.is_deleted" href="#GenomeGraphs.Graphs.is_deleted"><code>GenomeGraphs.Graphs.is_deleted</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Check if a SDGNode is a deleted.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.unsafe_sequence" href="#GenomeGraphs.Graphs.unsafe_sequence"><code>GenomeGraphs.Graphs.unsafe_sequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">unsafe_sequence(n::SDGNode{S}) where {S&lt;:BioSequence}</code></pre><p>Get the <strong>reference</strong> to a node&#39;s underlying sequence object.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>You get a reference to the node&#39;s sequence object - not a copy. So doing transformation operations (reverse_complement, setindex, etc.) to it will probably screw up the graph!</p></div></div></div></section><section><div><pre><code class="language-none">unsafe_sequence(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the <strong>reference</strong> to a node&#39;s underlying sequence object.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is unsafe, no checking of node id&#39;s occurs and you get a reference to the node&#39;s sequence object - not a copy, so doing transformation operations (reverse_complement, setindex, etc.) to it will probably screw up the graph!</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.check_node_id" href="#GenomeGraphs.Graphs.check_node_id"><code>GenomeGraphs.Graphs.check_node_id</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">check_node_id(sg::SequenceDistanceGraph, i::NodeID)</code></pre><p>The method responsible for checking that a node id used as input to a method that queries or edits a <code>SequenceDistanceGraph</code> is a sensible value.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.nodes" href="#GenomeGraphs.Graphs.nodes"><code>GenomeGraphs.Graphs.nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get a <strong>reference</strong> to the vector of nodes in a graph <code>sg</code>.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>It is a bad idea to edit this vector yourself unless you know what you are doing.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.node_unsafe" href="#GenomeGraphs.Graphs.node_unsafe"><code>GenomeGraphs.Graphs.node_unsafe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">node_unsafe(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a reference specific node from a sequence distance graph <code>sg</code> using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>node_unsafe</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the node for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to a graph&#39;s underlying <code>SDGNode</code>. NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is explicitly marked unsafe for a reason. It does <em>zero</em> checking of the value it is passed as the node id. It also makes use of the <code>@inbounds</code> macro. This makes it faster, but you need to be 100% sure that your code won&#39;t try to call this with a bad id. So it is not exported or recommended</p><ul><li>unless you really know what you&#39;re doing.</li></ul></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.node" href="#GenomeGraphs.Graphs.node"><code>GenomeGraphs.Graphs.node</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">node(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a <strong>reference</strong> to a specific node from a sequence distance graph <code>sg</code> using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>node</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the links for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to a graph&#39;s underlying <code>SDGNode</code>. NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div></div></section></article><h3 id="Graph-topology-1"><a class="docs-heading-anchor" href="#Graph-topology-1">Graph topology</a><a class="docs-heading-anchor-permalink" href="#Graph-topology-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.links" href="#GenomeGraphs.Graphs.links"><code>GenomeGraphs.Graphs.links</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get a <strong>reference</strong> to the vector of vectors of links in a graph <code>sg</code>.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>It is a bad idea to edit this vector yourself unless you know what you are doing.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.linksof_unsafe" href="#GenomeGraphs.Graphs.linksof_unsafe"><code>GenomeGraphs.Graphs.linksof_unsafe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">linksof_unsafe(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a <strong>reference</strong> to a vector storing all the links of a node in a <code>SequenceDistanceGraph</code>, the node is specified using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>links_unsafe</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the links for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to an underlying links vector that the <code>SequenceDistanceGraph</code> owns - NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is explicitly marked unsafe for a reason. It does <em>zero</em> checking of the value it is passed as the node id. It also makes use of the <code>@inbounds</code> macro. This makes it faster, but you need to be 100% sure that your code won&#39;t try to call this with a bad id. So it is not exported or recommended</p><ul><li>unless you really know what you&#39;re doing.</li></ul></div></div></div></section></article><h3 id="Graph-editing-and-manipulation-1"><a class="docs-heading-anchor" href="#Graph-editing-and-manipulation-1">Graph editing and manipulation</a><a class="docs-heading-anchor-permalink" href="#Graph-editing-and-manipulation-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.add_node!" href="#GenomeGraphs.Graphs.add_node!"><code>GenomeGraphs.Graphs.add_node!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">add_node!(sg::SequenceDistanceGraph{S}, n::SDGNode{S}) where {S&lt;:BioSequence}</code></pre><p>Add a node <code>n</code> to a graph <code>sg</code>.</p><p>Returns the node ID used to access the new node added in the graph.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>We don&#39;t enforce the sequence in the node is canonical here. We just trust that it is canonical.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>Adding a node to the graph does just that. After adding the node it still will not be linked to any other nodes.</p></div></div></div></section><section><div><p>add_node!(sg::SequenceDistanceGraph{S}, seq::BioSequence) where {S&lt;:BioSequence}</p><p>Add a sequence to a sequence distance graph as a node.</p><p>Returns the node ID used to access the new node added in the graph.</p><p>Can accept any sequence type and will attempt to coerce the input sequence to the type required by the graph.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>We don&#39;t enforce the sequence in the node is canonical here. We just trust that it is canonical.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>Adding a node to the graph does just that. After adding the node it still will not be linked to any other nodes.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.remove_node!" href="#GenomeGraphs.Graphs.remove_node!"><code>GenomeGraphs.Graphs.remove_node!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">remove_node!(sg::SequenceDistanceGraph{S}, n::NodeID) where {S&lt;:BioSequence}</code></pre><p>Remove a node from a sequence distance graph.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method can accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will end up deleting node 5.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Links involving this node will also be removed from the graph.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.add_link!" href="#GenomeGraphs.Graphs.add_link!"><code>GenomeGraphs.Graphs.add_link!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">add_link!(sg::SequenceDistanceGraph, source::NodeID, dest::NodeID, dist::Int)</code></pre><p>Construct a link between two node ends in a sequence Graph.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.remove_link!" href="#GenomeGraphs.Graphs.remove_link!"><code>GenomeGraphs.Graphs.remove_link!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">remove_link!(sg::SequenceDistanceGraph, src::NodeID, dst::NodeID)</code></pre><p>Remove a link between two nodes in a SequenceDistanceGraph. Returns a boolean indicating whether the removal was successful. Reasons this function would not return <code>true</code> include that the link didn&#39;t exist in the graph, and so could not be removed.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.disconnect_node!" href="#GenomeGraphs.Graphs.disconnect_node!"><code>GenomeGraphs.Graphs.disconnect_node!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Removes all the links in the collection from and to a given nodeID.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.Graphs.collapse_all_unitigs!" href="#GenomeGraphs.Graphs.collapse_all_unitigs!"><code>GenomeGraphs.Graphs.collapse_all_unitigs!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">collapse_all_unitigs!(unitigs::Vector{SequenceDistanceGraphPath{G}}, newnodes::Vector{NodeID}, sg::G, min_nodes::Integer, consume::Bool) where {G&lt;:SequenceDistanceGraph}</code></pre><p>Detects all of the trivial paths through the graph that define unitigs. Such paths are defined as a chain of nodes with only one neighbour. Each such simple paths through the graph can safely be collapsed into one larger node, which is what this function does.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method will edit <code>unitigs</code> so as after this method returns it contains All the paths that were detected as unitigs in the graph. This method also edits the <code>newnodes</code> vector, so as after this method returns it contains the IDs of all new nodes that were created as a result of collapsing the paths in <code>unitigs</code>.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Modifies the SequenceDistanceGraph <code>sg</code>.</p></div></div></div></section><section><div><pre><code class="language-none">collapse_all_unitigs!(sg::SequenceDistanceGraph, min_nodes::Integer, consume::Bool)</code></pre><p>Detects all of the trivial paths through the graph that define unitigs. Such paths are defined as a chain of nodes with only one neighbour. Each such simple paths through the graph can safely be collapsed into one larger node, which is what this function does.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Modifies the SequenceDistanceGraph <code>sg</code>.</p></div></div></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../man/guide/">« Guide</a><a class="docs-footer-nextpage" href="../MerTools/">MerTools submodule »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 17 January 2020 13:16">Friday 17 January 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
