<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GenomeGraphs.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomeGraphs.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href>Home</a><ul class="internal"><li><a class="toctext" href="#Introduction-1">Introduction</a></li><li><a class="toctext" href="#Framework-overview-1">Framework overview</a></li><li><a class="toctext" href="#TLDR;-Package-features-1">TLDR; Package features</a></li></ul></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="man/guide/">Guide</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Home</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Home</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="GenomeGraphs-1" href="#GenomeGraphs-1">GenomeGraphs</a></h1><p><em>A graph based genomics framework for the julia/BioJulia ecosystem</em></p><p><a href="https://github.com/BioJulia/GenomeGraphs/releases/latest"><img src="https://img.shields.io/github/release/BioJulia/GenomeGraphs.svg" alt="Latest release"/></a> <a href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-MIT-green.svg" alt="MIT license"/></a>  <a href="https://biojulia.github.io/GenomeGraphs/stable"><img src="https://img.shields.io/badge/docs-stable-blue.svg" alt="Stable documentation"/></a> <a href="http://www.repostatus.org/#wip"><img src="http://www.repostatus.org/badges/latest/wip.svg" alt="Pkg Status"/></a> <a href="https://gitter.im/BioJulia/GenomeGraphs"><img src="https://img.shields.io/gitter/room/BioJulia/GenomeGraphs.svg" alt="Chat"/></a></p><h2><a class="nav-anchor" id="Introduction-1" href="#Introduction-1">Introduction</a></h2><p>GenomeGraphs is designed to do one thing - provide a framework that makes it simple for a human to work with genome graphs from scripts or interactively.</p><p>Graphs are the core representation used by genome assemblers to represent genome sequence models constructed from reads. At the time of writing it is fair to say that until recently, their use has been limited to the internals of genome assemblers, which are often treated as black boxes that output a series of flattened sequences in FASTA format.</p><p>The use of graphs has increased in recent years thanks to the GFA file format and developments in genome variation graphs and sequence to graph mappers.</p><p>However, a lack of inter operation between graph-based tools, and limited tools for downstream graph-based analysis, contribute to a perceived complexity which maintains linear sequences and FASTA files as the typical unit of genomic sequence exchange.</p><p>Such flattening of graph representations within pipelines with multiple steps, that use different types of sequencing in an iterative fashion, produces ever-longer linear genome sequences through an information loss process.  As a result, genome assembly projects are prone to error propagation and difficult to reproduce and control.</p><p>To address these problems for the BioJulia ecosystem, GenomeGraphs provides a flexible framework for building and integrating information over genome graphs.</p><h2><a class="nav-anchor" id="Framework-overview-1" href="#Framework-overview-1">Framework overview</a></h2><p>This package implements a <code>SequenceDistanceGraph</code> type that represents sequences as nodes and the adjacency between sequences in links/edges. Rather than work directly with this graph data structure, you interact with a <code>WorkSpace</code>. A <code>WorkSpace</code> associates a <code>SequenceDistanceGraph</code> with raw sequencing reads,  sequence to graph mappings, and k-mer counts. The <code>WorkSpace</code> and the API  provides a working environment that enables you to project different kinds of  information over a graph, and navigate and analyse each node of a sequence graph.</p><p>Within the <code>WorkSpace</code>, you will find <code>DataStore</code> types permit random access to short, linked, and long read sequences stored on disk in <code>BioSequences.jl</code>&#39;s native bit encoding. Each datastore has an associated <code>Mapper</code> in the workspace that contains the output from mapping said reads onto the graph. <code>KmerCounts</code> allow you to compute k-mer coverage over the graph from sequencing data, enabling coverage analysis. Additional <code>DistanceGraph</code>s define alternative topologies over <code>SequenceDistanceGraph</code> nodes. They are typically used to represent longer range linkage information from various sequencing technologies.</p><p>Finally, a <code>NodeView</code> abstraction provides a proxy to a node, with methods to navigate a graph and access mapped data.</p><p>This framework is intended to allow the user to expore genome graphs interactively and to create processing methods for assembly or downstream analyses.</p><h2><a class="nav-anchor" id="TLDR;-Package-features-1" href="#TLDR;-Package-features-1">TLDR; Package features</a></h2><ul><li>Efficient on-disk (buffered) data stores for sequencing reads.</li><li>A simple graph data-structure for representing assembled genomes.</li><li>TODO: Transparent mapping of sequencing reads onto graphs.</li><li>TODO: Kmer counts and coverage projection over genome graph nodes.</li><li>Workspaces binding a genome graph, mapped sequences, kmer counts, and annotation.</li><li>De-novo genome assembly utilities:<ul><li>de-Bruijn graph construction, with tip-clipping &amp; bubble-popping.</li></ul></li></ul><footer><hr/><a class="next" href="man/guide/"><span class="direction">Next</span><span class="title">Guide</span></a></footer></article></body></html>
