<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Graphs · GenomeGraphs.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="GenomeGraphs.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">GenomeGraphs.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/guide/">Guide</a></li></ul></li><li><span class="tocitem">API</span><ul><li class="is-active"><a class="tocitem" href>Graphs</a><ul class="internal"><li><a class="tocitem" href="#Types-1"><span>Types</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Graphs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Graphs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/api/SequenceDistanceGraph.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API:-Graph-submodule-1"><a class="docs-heading-anchor" href="#API:-Graph-submodule-1">API: Graph submodule</a><a class="docs-heading-anchor-permalink" href="#API:-Graph-submodule-1" title="Permalink"></a></h1><h2 id="Types-1"><a class="docs-heading-anchor" href="#Types-1">Types</a><a class="docs-heading-anchor-permalink" href="#Types-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.SequenceDistanceGraph" href="#GenomeGraphs.SequenceDistanceGraph"><code>GenomeGraphs.SequenceDistanceGraph</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The SequenceDistanceGraph is a representation of a genome assembly. Sequences are contained in nodes, and the distances are represented by links.</p><p>A singe node represents a sequence <em>and</em> its reverse complement. Every node has a correlative ID starting from 1. For every node X in the graph, the negative ID -X is mapped to the reverse complement of X. This mapping is virtual: Only one node is stored in the graph. This is because every node has an orientaton: Each node has a positive end (+), and a negative end (-). So when a node is accessed with (or traversed by entering) the positive end the node yields the stored sequence. Conversely, when a node is accessed with (or traversed by entering) the negative end the node yelds the reverse complement of the stored sequence. In this way the positive end can be thought of as the sequence start, and the negative end can be thought of as the sequence end.</p><p>A single distance between two sequences is represented as a single link. Every link connects two node ends and contains a distance (they take the form <code>([+, -]n1, [+, -]n2, [+, -]dist)</code>). A link connects two node ends, and so the order of the signed nodes in the links does not change the link. If the distance in a link is negative, this represents an overlap between two sequences. These overlaps must be &quot;perfect overlaps&quot;.</p><p>This SequenceDistanceGraph type is only intended to be interacted with directly by developers and people who know what they are doing. Most tasks an end user wants to do to manipulate or query a graph can be </p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.SDG" href="#GenomeGraphs.SDG"><code>GenomeGraphs.SDG</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Shorthand for SequenceDistanceGraph</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.DistanceGraphLink" href="#GenomeGraphs.DistanceGraphLink"><code>GenomeGraphs.DistanceGraphLink</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Represents a single distance between two sequences in a SequenceDistanceGraph.</p></div></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>SDGNode</code>. Check Documenter&#39;s build log for details.</p></div></div><h3 id="Public-/-Safe-methods-1"><a class="docs-heading-anchor" href="#Public-/-Safe-methods-1">Public / Safe methods</a><a class="docs-heading-anchor-permalink" href="#Public-/-Safe-methods-1" title="Permalink"></a></h3><h4 id="Graph-Nodes-and-sequences-1"><a class="docs-heading-anchor" href="#Graph-Nodes-and-sequences-1">Graph Nodes and sequences</a><a class="docs-heading-anchor-permalink" href="#Graph-Nodes-and-sequences-1" title="Permalink"></a></h4><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.name" href="#GenomeGraphs.name"><code>GenomeGraphs.name</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the name of the graph. Defaults to the symbol :sdg.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.n_nodes" href="#GenomeGraphs.n_nodes"><code>GenomeGraphs.n_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the number of nodes in the sequence distance graph <code>sg</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.sequence" href="#GenomeGraphs.sequence"><code>GenomeGraphs.sequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">sequence(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the full sequence of a node in a sequence distance graph using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>sequence</code> accepts a NodeID that can be positive or negative. Nodes represent stretches of sequence in a canonical orientation, if you ask for for the sequence of say the third node, the positive node id 3 (which denotes traversing the third node in the forward direction), gives you the canonical sequence. If you use the negative ID -3 (which denotes traversing the third node in the reverse direction), you will get the reverse complement of the node&#39;s canonical (forward) sequence.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>It is safe to modify the returned sequence without screwing up your graph, yet thanks to BioSequences.jl&#39;s copy on write system for LongSequences, data copying will only occur if nessecery. You get the best of both worlds.</p></div></div></div></section><section><div><pre><code class="language-none">Get the sequence of the underlying SequenceDistanceGraph node.</code></pre></div></section></article><h4 id="Graph-Topology-1"><a class="docs-heading-anchor" href="#Graph-Topology-1">Graph Topology</a><a class="docs-heading-anchor-permalink" href="#Graph-Topology-1" title="Permalink"></a></h4><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.find_link" href="#GenomeGraphs.find_link"><code>GenomeGraphs.find_link</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_link(sg::SequenceDistanceGraph, src::NodeID, dst::NodeID)</code></pre><p>Find and return the link that exists between a source node and a destination node, using their correlative node ids.</p><p>In this instance, the IDs also denote the &quot;ends&quot; of a node, a link connects.</p><p>Recall that every node has an orientaton: Each node has a positive end (+), and a negative end (-). So when a node is accessed with (or traversed by entering) the positive end the node yields the stored sequence. Conversely, when a node is accessed with (or traversed by entering) the negative end the node yelds the reverse complement of the stored sequence.</p><p>Thus <code>find_link(sg, -5, 1)</code> means you want to find a link in the graph that allows you to exit the (-) end of node 5 (meaning you just traversed it in the  canonical + –&gt; - orientation), and enter the (+) end of node 1 (meaning you will also traverse node one in the canonical + –&gt; - orientation).</p><p>By contrast <code>find_link(sg, -5, -1)</code> means you want to find a link in the graph that allows you to exit the (-) end of node 5 - again having just traverse it in the + –&gt; - orientation, but want to enter node 1 through it&#39;s (-) end - and thus traversing node 1 in the + &lt;– - orientation, yielding the reverse complement of the canonical sequence of node 1.</p><p>If a link connecting the two node ends is present, you get it. If not, you get <code>nothing</code>. So make sure to check the output.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method is safe and public, but not reeeeaaallly intended for the end user, as they have to worry about links between &quot;ends&quot; of a node. In contrast to the rest of this framework, which rather has the user thinking in node centric terms. e.g. &quot;I am at node 5 which means I have traversed it forwards or I am at node -5, which means I have traversed node 5 backwards. From here I can visit node 1, -3, and 2, so I may go through nodes 1 and 2 forwards, and through 3 backwards. The methods <code>get_next_nodes</code> and <code>get_previous_nodes</code> are more user-friendly and follow this more node-centric mindset.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.forward_links" href="#GenomeGraphs.forward_links"><code>GenomeGraphs.forward_links</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">forward_links(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a vector of the links that are ahread of you, as you traverse node <code>n</code>, continuing forward in you present direction of travel (<code>n</code> can be a positive or negative node id).</p><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the links that allow you to leave node 5 in the canonical direction. If you used a negative ID such as -5, that means you are traversing node 5 in the non-canonical orientation, and so you will get the links that allow you to leave node 5 in the non-canonical direction.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.backward_links" href="#GenomeGraphs.backward_links"><code>GenomeGraphs.backward_links</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">backward_links(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a vector of the links that are behind of you, as you traverse node <code>n</code>, continuing forward in you present direction of travel (<code>n</code> can be a positive or negative node id).</p><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the links that would allow you to enter node 5 in the canonical direction. If you used a negative ID such as -5, that means you are traversing node 5 in the non-canonical orientation, and so you will get the links that would allow you to enter node 5 in the non-canonical direction.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.get_next_nodes" href="#GenomeGraphs.get_next_nodes"><code>GenomeGraphs.get_next_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_next_nodes(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the nodes you may visit next as you exit node <code>n</code>, maintaining your current direction of travel (<code>n</code> can be a positive or negative node ID).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the nodes you may visit as you leave node 5 in the canonical direction. If you use a negative ID, such as -5, this means you are traversing node 5 in the non-canonical orientation, and so you will get the nodes you may visit as you leave node 5 in the non-canonical direction.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The list of nodes returned are also signed to denote direction: Say you got <code>[1, -2, 3]</code> as a result of <code>get_next_nodes(sg, 5)</code>. That means as you leave node 5 after traversing it in the canonical direction, you may proceed to travel through nodes 1 and 3 in the canonical direction, and node 2 in the non-canonical direction.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.get_previous_nodes" href="#GenomeGraphs.get_previous_nodes"><code>GenomeGraphs.get_previous_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_previous_nodes(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the nodes you may have previously been on before you enterd node <code>n</code>, maintaining your current direction of travel (<code>n</code> can be a positive or negative node ID).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The node id can be positive or negative. For example if you use a positive ID, such as 5, this means you are traversing node 5 in the canonical orientation, and so you will get the nodes you may have visited prior to entering node 5 in the canonical direction. If you use a negative ID, such as -5, this means you are traversing node 5 in the non-canonical orientation, and so you will get the nodes you may have visited prior to entering node 5 in the non-canonical direction.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The list of nodes returned are also signed to denote direction: Say you got <code>[1, -2, 3]</code> as a result of <code>get_previous_nodes(sg, 5)</code>. That means prior to entering node 5 in the canonical direction, you may have traveled through nodes 1 and 3 in the canonical direction, and node 2 in the non-canonical direction.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.find_tip_nodes" href="#GenomeGraphs.find_tip_nodes"><code>GenomeGraphs.find_tip_nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_tip_nodes(sg::SequenceDistanceGraph, min_size::Integer)</code></pre><p>Get a set of IDs of all the nodes in the graph <code>sg</code> that count as &quot;tips&quot;.</p><p>Here a tip node is defined as a node that only has a single neighbouring node at one of it&#39;s ends, and no neighbouring nodes out of one of its other end.</p><p>The node&#39;s sequence must also be larger than <code>min_size</code> base pairs in length.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.find_tip_nodes!" href="#GenomeGraphs.find_tip_nodes!"><code>GenomeGraphs.find_tip_nodes!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_tip_nodes!(result::Set{NodeID}, sg::SequenceDistanceGraph, min_size::Integer)</code></pre><p>Get a set of IDs of all the nodes in the graph <code>sg</code> that count as &quot;tips&quot;.</p><p>Here a tip node is defined as a node that only has a single neighbouring node at one of it&#39;s ends, and no neighbouring nodes out of one of its other end.</p><p>The node&#39;s sequence must also be larger than <code>min_size</code> base pairs in length.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This method modifys a <code>result</code> input. So if you want to find tips in a graph repeatedly, you can resuse a <code>Set</code>, saving you some allocation costs.</p></div></div></div></section></article><h4 id="Graph-IO-1"><a class="docs-heading-anchor" href="#Graph-IO-1">Graph IO</a><a class="docs-heading-anchor-permalink" href="#Graph-IO-1" title="Permalink"></a></h4><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>write_to_gfa1</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.load_from_gfa1!" href="#GenomeGraphs.load_from_gfa1!"><code>GenomeGraphs.load_from_gfa1!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">load_from_gfa1!(sg::SequenceDistanceGraph{S}, gfafile::AbstractString, fafile::AbstractString) where {S&lt;:BioSequence}</code></pre><p>Load a graph from a GFAv1 formatted file, and associated FASTA file of node sequences.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The GFA format permits storing sequences of the graph nodes in a seperate fasta file, instead of in the GFA file. This is so as the sequences of the graph nodes can be easily fed into other tools that typically accept FASTA files as input. Many assemblers also output a GFA + FASTA combo.  Therefore, this method asks for the filepath of a GFAv1 file, as well as a filepath to a FASTA formatted file. This method reads the node sequences from the FASTA file, before getting the links between nodes from the GFAv1 file.   </p></div></div></div></section></article><h3 id="Internal-/-Unsafe-methods-1"><a class="docs-heading-anchor" href="#Internal-/-Unsafe-methods-1">Internal / Unsafe methods</a><a class="docs-heading-anchor-permalink" href="#Internal-/-Unsafe-methods-1" title="Permalink"></a></h3><h4 id="Accessors-and-property-queries-1"><a class="docs-heading-anchor" href="#Accessors-and-property-queries-1">Accessors and property queries</a><a class="docs-heading-anchor-permalink" href="#Accessors-and-property-queries-1" title="Permalink"></a></h4><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.check_node_id" href="#GenomeGraphs.check_node_id"><code>GenomeGraphs.check_node_id</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">check_node_id(sg::SequenceDistanceGraph, i::NodeID)</code></pre><p>The method responsible for checking that a node id used as input to a method that queries or edits a <code>SequenceDistanceGraph</code> is a sensible value.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.nodes" href="#GenomeGraphs.nodes"><code>GenomeGraphs.nodes</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get a <strong>reference</strong> to the vector of nodes in a graph <code>sg</code>.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>It is a bad idea to edit this vector yourself unless you know what you are doing.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.node_unsafe" href="#GenomeGraphs.node_unsafe"><code>GenomeGraphs.node_unsafe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">node_unsafe(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a refernece specific node from a sequence distance graph <code>sg</code> using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>node_unsafe</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the node for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to a graph&#39;s underlying <code>SDGNode</code>. NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is explicitly marked unsafe for a reason. It does <em>zero</em> checking of the value it is passed as the node id. It also makes use of the <code>@inbounds</code> macro. This makes it faster, but you need to be 100% sure that your code won&#39;t try to call this with a bad id. So it is not exported or recommended</p><ul><li>unless you really know what you&#39;re doing.</li></ul></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.sequence_unsafe" href="#GenomeGraphs.sequence_unsafe"><code>GenomeGraphs.sequence_unsafe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">sequence_unsafe(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get the reference to a node&#39;s underlying sequence object.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is unsafe, no checking of node id&#39;s occurs and you get a reference to the node&#39;s sequence object - not a copy, so doing transformation operations (reverse_complement, setindex, etc.) to it will probably screw up the graph!</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.links" href="#GenomeGraphs.links"><code>GenomeGraphs.links</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get a <strong>reference</strong> to the vector of vectors of links in a graph <code>sg</code>.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>It is a bad idea to edit this vector yourself unless you know what you are doing.</p></div></div></div></section><section><div><pre><code class="language-none">links(sg::SequenceGraph, n::NodeID)

Get a **reference** to a vector storing all the links of a node in a
`SequenceDistanceGraph`, the node is specified using its correlative node id `n`.</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>links</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the links for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to an underlying links vector that the <code>SequenceDistanceGraph</code> owns - NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.links_unsafe" href="#GenomeGraphs.links_unsafe"><code>GenomeGraphs.links_unsafe</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">links_unsafe(sg::SequenceDistanceGraph, n::NodeID)</code></pre><p>Get a <strong>reference</strong> to a vector storing all the links of a node in a <code>SequenceDistanceGraph</code>, the node is specified using its correlative node id <code>n</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>links_unsafe</code> accepts a NodeID that can be positive or negative. E.g. providing either 5 or -5 both mean node 5 in a graph, and so you will get the links for node 5.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method returns a <strong>reference</strong> to an underlying links vector that the <code>SequenceDistanceGraph</code> owns - NOT a copy! Messing with this will screw up your graph. So this method is not recommended or exported - unless you really know what you&#39;re doing.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This method is explicitly marked unsafe for a reason. It does <em>zero</em> checking of the value it is passed as the node id. It also makes use of the <code>@inbounds</code> macro. This makes it faster, but you need to be 100% sure that your code won&#39;t try to call this with a bad id. So it is not exported or recommended</p><ul><li>unless you really know what you&#39;re doing.</li></ul></div></div></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../man/guide/">« Guide</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 20 December 2019 17:16">Friday 20 December 2019</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
