<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>The Sequence Distance Graph (SDG) · GenomeGraphs.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="GenomeGraphs.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">GenomeGraphs.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/guide/">Guide</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../../api/SequenceDistanceGraph/">Graphs</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>The Sequence Distance Graph (SDG)</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>The Sequence Distance Graph (SDG)</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/types/graphs.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="The-Sequence-Distance-Graph-(SDG)-1"><a class="docs-heading-anchor" href="#The-Sequence-Distance-Graph-(SDG)-1">The Sequence Distance Graph (SDG)</a><a class="docs-heading-anchor-permalink" href="#The-Sequence-Distance-Graph-(SDG)-1" title="Permalink"></a></h1><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>SequenceDistanceGraph</code>. Check Documenter&#39;s build log for details.</p></div></div><h1 id="Querying-an-SDG-for-basic-properties-1"><a class="docs-heading-anchor" href="#Querying-an-SDG-for-basic-properties-1">Querying an SDG for basic properties</a><a class="docs-heading-anchor-permalink" href="#Querying-an-SDG-for-basic-properties-1" title="Permalink"></a></h1><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>nodes</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>n_nodes</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="GenomeGraphs.each_node_id" href="#GenomeGraphs.each_node_id"><code>GenomeGraphs.each_node_id</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Iterate over every node ID in the sequence distance graph <code>sg</code>.</p></div></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>links</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>node</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>sequence</code>. Check Documenter&#39;s build log for details.</p></div></div><h1 id="Manually-editing-an-SDG-by-manipulating-nodes-and-links-1"><a class="docs-heading-anchor" href="#Manually-editing-an-SDG-by-manipulating-nodes-and-links-1">Manually editing an SDG by manipulating nodes and links</a><a class="docs-heading-anchor-permalink" href="#Manually-editing-an-SDG-by-manipulating-nodes-and-links-1" title="Permalink"></a></h1><p>It is not recommended you do this if you are a high level user. However these small editing operations are required for developers. If you find yourself needing these methods, you will have to explicitly import them, as they are not exported from the module.</p><p>If you find yourself wanting to edit the graph manually, it&#39;s a good idea to ask the package authors listed in the <code>Project.toml</code> or <code>.github/CODEOWNERS</code>.</p><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>add_node!</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>remove_node!</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>add_link!</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>remove_link!</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>disconnect_node!</code>. Check Documenter&#39;s build log for details.</p></div></div><p>#Code example for testing the error correction and simplifying the graph</p><p>In the example below we generate a toy graph from 4-mers which contain a bubble. As for now, we make use of randomly generated coverage information for</p><pre><code class="language-none">using BioSequences
using BioSequenceGraphs

kmerlist = Vector{DNAKmer{4}}([DNAKmer{4}(&quot;ATAC&quot;),DNAKmer{4}(&quot;TACG&quot;),DNAKmer{4}(&quot;TACC&quot;),DNAKmer{4}(&quot;ACGA&quot;),DNAKmer{4}(&quot;CGAA&quot;),DNAKmer{4}(&quot;GAAT&quot;),DNAKmer{4}(&quot;AATC&quot;),DNAKmer{4}(&quot;ACCA&quot;),DNAKmer{4}(&quot;CCAA&quot;),DNAKmer{4}(&quot;CAAT&quot;)])
sdg = SequenceDistanceGraph(kmerlist,true)</code></pre><p>The new SequenceDistanceGraph constructor takes as input a list of kmerlist (not necessarily sorted or in canonical form) and a boolean for whether to do error correction or not. Then by using the coverage information generated randomly, simplifies the graph. The resulting SDG before and after the error correction steps are as follows:</p><p>***Before:***</p><pre><code class="language-none">SequenceDistanceGraph{BioSequence{DNAAlphabet{2}}}(SDGNode{BioSequence{DNAAlphabet{2}}}[SDGNode{BioSequence{DNAAlphabet{2}}}(AATC, false), SDGNode{BioSequence{DNAAlphabet{2}}}(ATAC, false), SDGNode{BioSequence{DNAAlphabet{2}}}(ATTCGTA, false), SDGNode{BioSequence{DNAAlphabet{2}}}(ATTGGTA, false)], Array{DistanceGraphLink,1}[[DistanceGraphLink(1, 3, -3), DistanceGraphLink(1, 4, -3)], [DistanceGraphLink(-2, -4, -3), DistanceGraphLink(-2, -3, -3)], [DistanceGraphLink(3, 1, -3), DistanceGraphLink(-3, -2, -3)], [DistanceGraphLink(4, 1, -3), DistanceGraphLink(-4, -2, -3)]])</code></pre><p>***After:***</p><pre><code class="language-none">SequenceDistanceGraph{BioSequence{DNAAlphabet{2}}}(SDGNode{BioSequence{DNAAlphabet{2}}}[SDGNode{BioSequence{DNAAlphabet{2}}}(GATTCGTAC, false)], Array{DistanceGraphLink,1}[[]])</code></pre><p>So the algorithm first generates a sdg with collapsing all simple paths using the initial kmer list. Then at the next stage the low covered bubble branch is removed and path collapsing is applied one more time. This results in a sdg with a whole single node, a perfect unitig.</p><p>##TODO:</p><p>Implement the sdg constructor to take as input a list of reads instead of kmers and generate all kmer + coverage information from them.</p></article></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 20 December 2019 02:57">Friday 20 December 2019</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
