<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>- · GenomeGraphs.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomeGraphs.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="../man/guide/">Guide</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>-</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/DeBruijnGraph.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>-</span><a class="fa fa-bars" href="#"></a></div></header><h2><a class="nav-anchor" id="de-Bruijn-Graph-type-1" href="#de-Bruijn-Graph-type-1">de Bruijn Graph type</a></h2><p>A fundamental approach for de-novo gene assembly is to use make use of de Bruijn graphs. The graph is used to represent fragments of reads (mostly starting with kmers) as vertices and overlaps between these fragments as edges. <code>DeBruijnGraph</code> type is a special type of <code>SequenceGraph</code>. It is also made up of two fields:</p><ul><li>nodes</li><li>links</li></ul><pre><code class="language-none">struct DeBruijnGraph
    nodes::Vector{SequenceGraphNode}
    links::Vector{Vector{SequenceGraphLink}}
end</code></pre><p>We initialize a DeBruijnGraph using the <code>deBruijn_constructor</code> function. This is mainly due to the fact that arbitrary links between two vertices are not allowed in the de Bruijn graph formalism. The constructor receives as input a list of kmers and generates the deBruijn_Graph where each kmer is a unique vertex and each overlap  of length <span>$k-1$</span> is represented with an edge.</p><pre><code class="language-none">kmer_vector = generate_random_kmers(DNA,4,10)
dbg = deBruijn_constructor(kmer_vector)

DeBruijnGraph(SequenceGraphNode[SequenceGraphNode{Kmer{DNA,4}}(CGCC, true), SequenceGraphNode{Kmer{DNA,4}}(TCTG, true), SequenceGraphNode{Kmer{DNA,4}}(TGTG, true), SequenceGraphNode{Kmer{DNA,4}}(GAAG, true), SequenceGraphNode{Kmer{DNA,4}}(GGCA, true), SequenceGraphNode{Kmer{DNA,4}}(ACGA, true), SequenceGraphNode{Kmer{DNA,4}}(CGCT, true), SequenceGraphNode{Kmer{DNA,4}}(TCTC, true), SequenceGraphNode{Kmer{DNA,4}}(TACG, true), SequenceGraphNode{Kmer{DNA,4}}(GCAT, true)], Array{SequenceGraphLink,1}[[], [], [], [], [SequenceGraphLink(-5, 10, 1)], [], [], [], [SequenceGraphLink(-9, 6, 1)], []])</code></pre><p>We can use the fastq readers to work on real data. The example below generates the set of unique kmers from 5-long reads. The kmer size is set to 15.</p><pre><code class="language-none">r = FASTQ.Reader(open(&quot;URnano_ecoli.fastq&quot;, &quot;r&quot;))
Ecoli_reads = Set{BioSequence{DNAAlphabet{4}}}()

## get first 5 reads
for i in 1:5
    next_seq = iterate(r)
    seq = sequence(next_seq[1])
    next_seq = iterate(r,next_seq[2])
    push!(Ecoli_reads,seq)
end
kmers = new_extract_canonical_kmers(Ecoli_reads,15)</code></pre><p>The output is a set of kmers in their canonical form:</p><p>Set(Kmer{DNA,15}[AAAATATCTCGTTTT, TAAACCAGTCGCCGC, TTCGACATTACCCAG, CGCCCTGCCAGCAGT, TAATATTGTTCCATT, TGGTAATGGTCACAG, AAAAATTAAGCAGGA, ATATAAGTTATATCA, GCCCGATCTGTCTCC, GATTTCTCCGGGCCA  …  TGAGCGATTGCCTGA, CGGAGCAGCAGTGTC, AAAATCGTACATACC, GTCGCCTGATGCCTG, GTCAGCGAACCTTCC, CGCCGCTCACCGCCG, TGGATGAACGTTCAT, AATATGTCACAATTT, ATGCGATAGCAGGGG, GTAGAAAGCTCGTGG, CGATTGGTTTAAGAC])</p><p>Example data for checking node merging :</p><pre><code class="language-none">kmer_vector2 = Vector([DNAKmer{4}(&quot;ATTC&quot;),DNAKmer{4}(&quot;TTCG&quot;),DNAKmer{4}(&quot;TCGT&quot;),
        DNAKmer{4}(&quot;AATC&quot;),DNAKmer{4}(&quot;AATG&quot;),DNAKmer{4}(&quot;CGTA&quot;),DNAKmer{4}(&quot;CGTC&quot;)])</code></pre><p>DeBruijnGraph for the above kmers:</p><pre><code class="language-none">DeBruijnGraph(Dict{Int64,SequenceGraphNode}(7=&gt;SequenceGraphNode{Kmer{DNA,4}}(ACGA, true),4=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTC, true),2=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATG, true),3=&gt;SequenceGraphNode{Kmer{DNA,4}}(ATTC, true),5=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTA, true),6=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGAA, true),1=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATC, true)), Dict(7=&gt;[SequenceGraphLink(-7, 6, -3)],4=&gt;[SequenceGraphLink(4, 7, -3)],2=&gt;[],3=&gt;[SequenceGraphLink(3, 1, -3), SequenceGraphLink(3, 2, -3)],5=&gt;[SequenceGraphLink(5, 7, -3)],6=&gt;[SequenceGraphLink(-6, -3, -3)],1=&gt;[]), 4)</code></pre><p>Then we apply node merging:</p><pre><code class="language-none">DeBruijnGraph(Dict{Int64,SequenceGraphNode}(7=&gt;SequenceGraphNode{BioSequence{DNAAlphabet{4}}}(ACGAAT, true),4=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTC, true),2=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATG, true),5=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTA, true),1=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATC, true)), Dict(7=&gt;[SequenceGraphLink(-7, 6, -3), SequenceGraphLink(-7, 1, -3), SequenceGraphLink(-7, 2, -3)],4=&gt;[SequenceGraphLink(4, 7, -3)],2=&gt;[],5=&gt;[SequenceGraphLink(5, 7, -3)],1=&gt;[]), 4)</code></pre><p>So the nodes with nodeID 6 and 3 are collapsed into node 7 and we can see that both outgoing edges of 3 are given to node 7.</p><footer><hr/></footer></article></body></html>
