<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>- · GenomeGraphs.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="GenomeGraphs.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">GenomeGraphs.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../man/guide/">Guide</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../api/Graphs/">Graphs submodule</a></li><li><a class="tocitem" href="../api/MerTools/">MerTools submodule</a></li><li><a class="tocitem" href="../api/GraphIndexes/">GraphIndexes submodule</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>-</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>-</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/GenomeGraphs.jl/blob/master/docs/src/DeBruijnGraph.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h2 id="de-Bruijn-Graph-type-1"><a class="docs-heading-anchor" href="#de-Bruijn-Graph-type-1">de Bruijn Graph type</a><a class="docs-heading-anchor-permalink" href="#de-Bruijn-Graph-type-1" title="Permalink"></a></h2><p>A fundamental approach for de-novo gene assembly is to use make use of de Bruijn graphs. The graph is used to represent fragments of reads (mostly starting with kmers) as vertices and overlaps between these fragments as edges. <code>DeBruijnGraph</code> type is a special type of <code>SequenceGraph</code>. It is also made up of two fields:</p><ul><li>nodes</li><li>links</li></ul><pre><code class="language-none">struct DeBruijnGraph
    nodes::Vector{SequenceGraphNode}
    links::Vector{Vector{SequenceGraphLink}}
end</code></pre><p>We initialize a DeBruijnGraph using the <code>deBruijn_constructor</code> function. This is mainly due to the fact that arbitrary links between two vertices are not allowed in the de Bruijn graph formalism. The constructor receives as input a list of kmers and generates the deBruijn_Graph where each kmer is a unique vertex and each overlap  of length <span>$k-1$</span> is represented with an edge.</p><pre><code class="language-none">kmer_vector = generate_random_kmers(DNA,4,10)
dbg = deBruijn_constructor(kmer_vector)

DeBruijnGraph(SequenceGraphNode[SequenceGraphNode{Kmer{DNA,4}}(CGCC, true), SequenceGraphNode{Kmer{DNA,4}}(TCTG, true), SequenceGraphNode{Kmer{DNA,4}}(TGTG, true), SequenceGraphNode{Kmer{DNA,4}}(GAAG, true), SequenceGraphNode{Kmer{DNA,4}}(GGCA, true), SequenceGraphNode{Kmer{DNA,4}}(ACGA, true), SequenceGraphNode{Kmer{DNA,4}}(CGCT, true), SequenceGraphNode{Kmer{DNA,4}}(TCTC, true), SequenceGraphNode{Kmer{DNA,4}}(TACG, true), SequenceGraphNode{Kmer{DNA,4}}(GCAT, true)], Array{SequenceGraphLink,1}[[], [], [], [], [SequenceGraphLink(-5, 10, 1)], [], [], [], [SequenceGraphLink(-9, 6, 1)], []])</code></pre><p>We can use the fastq readers to work on real data. The example below generates the set of unique kmers from 5-long reads. The kmer size is set to 15.</p><pre><code class="language-none">r = FASTQ.Reader(open(&quot;URnano_ecoli.fastq&quot;, &quot;r&quot;))
Ecoli_reads = Vector{BioSequence{DNAAlphabet{4}}}()

## get first 5 reads
for i in 1:5
    next_seq = iterate(r)
    seq = sequence(next_seq[1])
    next_seq = iterate(r,next_seq[2])
    push!(Ecoli_reads,seq)
end
kmers = new_extract_canonical_kmers(Ecoli_reads,15)</code></pre><p>The output is a set of kmers in their canonical form:</p><p>Set(Kmer{DNA,15}[AAAATATCTCGTTTT, TAAACCAGTCGCCGC, TTCGACATTACCCAG, CGCCCTGCCAGCAGT, TAATATTGTTCCATT, TGGTAATGGTCACAG, AAAAATTAAGCAGGA, ATATAAGTTATATCA, GCCCGATCTGTCTCC, GATTTCTCCGGGCCA  …  TGAGCGATTGCCTGA, CGGAGCAGCAGTGTC, AAAATCGTACATACC, GTCGCCTGATGCCTG, GTCAGCGAACCTTCC, CGCCGCTCACCGCCG, TGGATGAACGTTCAT, AATATGTCACAATTT, ATGCGATAGCAGGGG, GTAGAAAGCTCGTGG, CGATTGGTTTAAGAC])</p><p>Example data for checking node merging :</p><pre><code class="language-none">kmer_vector2 = Vector([DNAKmer{4}(&quot;ATTC&quot;),DNAKmer{4}(&quot;TTCG&quot;),DNAKmer{4}(&quot;TCGT&quot;),
        DNAKmer{4}(&quot;AATC&quot;),DNAKmer{4}(&quot;AATG&quot;),DNAKmer{4}(&quot;CGTA&quot;),DNAKmer{4}(&quot;CGTC&quot;)])</code></pre><p>DeBruijnGraph for the above kmers:</p><pre><code class="language-none">DeBruijnGraph(Dict{Int64,SequenceGraphNode}(7=&gt;SequenceGraphNode{Kmer{DNA,4}}(ACGA, true),4=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTC, true),2=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATG, true),3=&gt;SequenceGraphNode{Kmer{DNA,4}}(ATTC, true),5=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTA, true),6=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGAA, true),1=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATC, true)), Dict(7=&gt;[SequenceGraphLink(-7, 6, -3)],4=&gt;[SequenceGraphLink(4, 7, -3)],2=&gt;[],3=&gt;[SequenceGraphLink(3, 1, -3), SequenceGraphLink(3, 2, -3)],5=&gt;[SequenceGraphLink(5, 7, -3)],6=&gt;[SequenceGraphLink(-6, -3, -3)],1=&gt;[]), 4)</code></pre><p>Then we apply node merging:</p><pre><code class="language-none">DeBruijnGraph(Dict{Int64,SequenceGraphNode}(7=&gt;SequenceGraphNode{BioSequence{DNAAlphabet{4}}}(ACGAAT, true),4=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTC, true),2=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATG, true),5=&gt;SequenceGraphNode{Kmer{DNA,4}}(CGTA, true),1=&gt;SequenceGraphNode{Kmer{DNA,4}}(AATC, true)), Dict(7=&gt;[SequenceGraphLink(-7, 6, -3), SequenceGraphLink(-7, 1, -3), SequenceGraphLink(-7, 2, -3)],4=&gt;[SequenceGraphLink(4, 7, -3)],2=&gt;[],5=&gt;[SequenceGraphLink(5, 7, -3)],1=&gt;[]), 4)</code></pre><p>So the nodes with nodeID 6 and 3 are collapsed into node 7 and we can see that both outgoing edges of 3 are given to node 7.</p><h4 id="Saving-to-gfa-1"><a class="docs-heading-anchor" href="#Saving-to-gfa-1">Saving to gfa</a><a class="docs-heading-anchor-permalink" href="#Saving-to-gfa-1" title="Permalink"></a></h4><p>Below is an example code for saving a SequenceDistanceGraph to GFA file. This allows us to load and visualize the graph using graph visualization tools.</p><pre><code class="language-none"># A helpful function to let me run Bandage to visualize graphs.
const BANDAGE_BIN = &quot;/Applications/Bandage.app/Contents/MacOS/Bandage&quot;
function draw_graph(gr)
    filename = tempname()
    BioSequenceGraphs.dump_to_gfa1(gr, filename)
    run(`$BANDAGE_BIN image $filename.gfa $filename.png --height 500`)
    run(`open $filename.png`)
end
function show_graph(gr)
    filename = tempname()
    BioSequenceGraphs.dump_to_gfa1(gr, filename)
    run(`$BANDAGE_BIN load $filename.gfa`)
end</code></pre></article></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 4 March 2020 14:20">Wednesday 4 March 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
